File: files/CH-029.py
import pandas as pd

input = pd.read_excel("CH-029 Identifying Customers Staple Products.xlsx", sheet_name="Sheet1", usecols="B:E", skiprows=1)
test = pd.read_excel("CH-029 Identifying Customers Staple Products.xlsx", sheet_name="Sheet1", usecols="I:J", skiprows=1, nrows = 4)

result = input.groupby(["Customer ID", "Product"]).agg(total_quantity=("Quantity", "sum")).reset_index()
result["rank"] = result.groupby("Customer ID")["total_quantity"].rank(ascending=False)
result["lowest_rank"] = result.groupby("Customer ID")["rank"].transform("min")
result = result[result["rank"] == result["lowest_rank"]]
result = result.groupby("Customer ID").agg({"Product": lambda x: ",".join(sorted(x))}).reset_index()
result = result.rename(columns={"Product": "Most Purchased PRODUCT"})

print(result["Most Purchased PRODUCT"].equals(test["Most Purchased PRODUCT"])) # True

File: files/000-099/029/CH-029.R
library(tidyverse)
library(readxl)

input = read_excel("files/CH-029 Identifying Customers Staple Products.xlsx", range = "B2:E36")
test  = read_excel("files/CH-029 Identifying Customers Staple Products.xlsx", range = "I2:J6")

result = input %>%
  summarise(total_quantity = sum(Quantity), .by = c("Customer ID", "Product")) %>%
  group_by(`Customer ID`) %>%
  mutate(rank = rank(-total_quantity),
         lowest_rank = min(rank)) %>%
  filter(rank == lowest_rank) %>%
  summarise(`Most Purchased PRODUCT` = paste0(sort(Product), collapse = ","))

identical(result$`Most Purchased PRODUCT`, test$`Most Purchased PRODUCT`)
# [1] TRUE
